python/gtsam/utils/plot.py

```diff
@@ -87,7 +88,7 @@ def plot_covariance_ellipse_3d(axes,
         n: Defines the granularity of the ellipse. Higher values indicate finer ellipses.
         alpha: Transparency value for the plotted surface in the range [0, 1].
     """
-    k = 11.82
+    k = np.sqrt(14.157)
```

Comment:
above = 14.157 and here it is sqrt(14.157)

Did you take a look at the PR #1063 as I suggested?

Reply:
I took a look at PR #1063 but am not sure how it is relevant. Perhaps I am missing something? I am new to contributing to open source projects so perhaps there was a file or something I needed to look at. As far as I can tell, the discussion there is regarding removing a negative sign in equations 5.3 and 5.6 in the `math.pdf` document.

As for the `k` values I chose, they come from the Chi Squared distribution's p-table for different dimensions. E.g., for a single dimension, `k = 9` corresponds to 3 standard deviations, for 2D we have `k = 11.820`, for 3D we have `k = 14.157`, and for 4D `k = 16.251` (all values taken from the table on page 366 of the Maybeck Vol 1 textbook). Note that I am choosing `k` to be the variance rather than the standard deviation, which is why you see the square root. I chose to write the comments as I did simply because I wanted to keep them as similar as possible to the previous comments pointing out the `k` values.

The most useful online reference I could find for choosing these `k` values can be found [here](https://www.visiondummy.com/2014/04/draw-error-ellipse-representing-covariance-matrix/) and the Maybeck textbook can be found [here](https://ai.unist.ac.kr/~chiu/robot/papers/Maybeck_P.S.-Stochastic_Models,_Estimation_and_Control_1979.pdf).

Please let me know if I can answer any other questions.

Thanks!

Reply:
Take a look at my note at the top of [plot.py](https://github.com/borglab/gtsam/blob/11c5bb97666877fa1e6a8f50e07ed3c9127ca539/python/gtsam/utils/plot.py). I added some python code for 2D.

1. In 2D I chose to leave the k=5, because it is consistent with what you said in the comments, and it is referenced in the documentation.
2. I don't think we should change the semantics of k in the middle of the game.

Given the note and 1 and 2 above, do you agree that code was already good as it was? You would then only have to bring 3D in line.

Reply:
Just for reference, we can easily calculate that in python with scipy (even though I guess we won't include that as a dependency, we can put the code in the comment to regenerate the values):

```python

def pct_to_sigma(pct, dof):
    return np.sqrt(scipy.stats.chi2.ppf(pct / 100., df=dof))

def sigma_to_pct(sigma, dof):
    return scipy.stats.chi2.cdf(sigma**2, df=dof) * 100.

for dim in range(0, 4):
    print("{}D".format(dim), end="")
    for n_sigma in range(1, 6):
        if dim == 0: print("\t    {}    ".format(n_sigma), end="")
        else: print("\t{:.5f}".format(sigma_to_pct(n_sigma, dim)), end="")
    print()

```

```
0D	    1    	    2    	    3    	    4    	    5    
1D	68.26895	95.44997	99.73002	99.99367	99.99994
2D	39.34693	86.46647	98.88910	99.96645	99.99963
3D	19.87480	73.85359	97.07091	99.88660	99.99846

```

Reply:
Cool, this code is more general than the one I put in plot.py (which I assume you looked at) and it agrees with it in the 2D case. So I propose:
- undo the changes to 2D plotting
- we keep k=5 for 2D
- k == sigma, not variance
- use k = 5 for 3D as well, which is 99.99846 of all probability.
- make sure it multiplies sigmas, not variances.

Reply:
> Just for reference, we can easily calculate that in python with scipy (even though I guess we won't include that as a dependency, we can put the code in the comment to regenerate the values):

Actually, maybe we should include this code, and comment on some of its output values, just like @senselessDev did below. I think that it makes sense for users to specify an integer number of standard deviations, and we can document in 2D and 3D how much percentage 1,2,3,4,5 stddevs corresponds to. k can be an input argument defaulting to 5. But we could provide another optional argument, `percentage` that, when given, calculates k using `pct_to_sigma`.

> 0D	    1    	    2    	    3    	    4    	    5    
> 1D	68.26895	95.44997	99.73002	99.99367	99.99994
> 2D	39.34693	86.46647	98.88910	99.96645	99.99963
> 3D	19.87480	73.85359	97.07091	99.88660	99.99846
> ```